src/Workspaces/Core/Portable/Classification/IRemoteSemanticClassificationService.cs

```diff
@@ -82,19 +87,14 @@ private static SerializableClassifiedSpans Dehydrate(ImmutableArray<ClassifiedSp
                 classificationTriples.Add(textSpan.Length);
             }
 
-            return new SerializableClassifiedSpans
-            {
-                ClassificationTypes = classificationTypes,
-                ClassificationTriples = classificationTriples,
-            };
+            return new SerializableClassifiedSpans(
+                classificationTypes.ToImmutableAndClear(),
+                classificationTriples.ToImmutableAndClear());
         }
 
-        internal void Rehydrate(ArrayBuilder<ClassifiedSpan> classifiedSpans)
+        internal void Rehydrate(SegmentedList<ClassifiedSpan> classifiedSpans)
         {
-            Contract.ThrowIfNull(ClassificationTypes);
-            Contract.ThrowIfNull(ClassificationTriples);
-
-            for (var i = 0; i < ClassificationTriples.Count; i += 3)
+            for (int i = 0, n = ClassificationTriples.Length; i < n; i += 3)
```

Comment:
>n = ClassificationTriples.Length; i < n; [](http://example.com/codeflow?start=28&length=40)

learning question for me. I've seen you do this a couple times, is it primarily in cases where 'n' comes from a class member because the compiler can't infer that it doesn't change?

Reply:
correct.  this ist just hoisting out the potentially variable value to a known fixed value.